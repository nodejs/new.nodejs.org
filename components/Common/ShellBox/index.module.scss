.shellBox {
  background-color: var(--black10);
  border-radius: 0.4rem;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  font-family: var(--mono);
  position: relative;

  code {
    color: var(--pink5);
    font-family: inherit;
    margin: 1em;
    overflow-x: hidden;

    &:hover {
      overflow-x: auto;
    }

    &::-webkit-scrollbar {
      height: 0.5em;
    }

    &::-webkit-scrollbar,
    &::-webkit-scrollbar-thumb {
      border-radius: 4px;
      overflow: visible;
    }

    &::-webkit-scrollbar-thumb {
      background: rgba(0, 0, 0, 0.2);
    }

    > span.function {
      color: var(--warning5);
    }
  }

  .top {
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    span,
    button {
      align-items: center;
      display: inherit;
      font-size: var(--font-size-code);
      height: 23px;
      justify-content: center;
      width: 86px;
    }

    span {
      background-color: var(--black3);
      border-radius: 0 0 0.3rem 0.3rem;
      color: var(--black9);
    }

    button {
      background-color: var(--brand);
      border-radius: 0 0.3rem 0.3rem 0.3rem;
      border-width: 0;
      cursor: pointer;
      i {
        padding: 0;
      }
    }
  }
}
