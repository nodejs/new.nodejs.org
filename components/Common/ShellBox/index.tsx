import { useRef } from 'react';
import { FormattedMessage } from 'react-intl';
import styles from './index.module.scss';
import { useCopyToClipboard } from '../../../hooks/useCopyToClipboard';
import type { FC, PropsWithChildren, MouseEvent, ReactNode } from 'react';

type ShellBoxProps = {
  children: string | ReactNode;
  textToCopy?: string;
};

const ShellBox: FC<PropsWithChildren<ShellBoxProps>> = ({
  children,
  textToCopy,
}: PropsWithChildren<ShellBoxProps>) => {
  const [copied, copyText] = useCopyToClipboard();

  const shellBoxRef = useRef<HTMLElement>(null);

  const handleCopyCode = async (event: MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    // By default we want to use the textToCopy props but if it's absent
    // we allow the user to copy by getting the inner HTML content of the Element
    const _textToCopy = textToCopy || shellBoxRef.current?.innerHTML || '';

    await copyText(_textToCopy.replace('$', ''));
  };

  return (
    <pre className={styles.shellBox}>
      <div className={styles.top}>
        <span>SHELL</span>
        <button type="button" onClick={handleCopyCode}>
          <FormattedMessage
            id="components.common.shellBox.copy"
            values={{ copied }}
          />
        </button>
      </div>
      <code ref={shellBoxRef}>{children}</code>
    </pre>
  );
};

export default ShellBox;
