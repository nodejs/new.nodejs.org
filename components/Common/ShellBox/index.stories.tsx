import ShellBox from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof ShellBox>;
type Meta = MetaObj<typeof ShellBox>;

export const Default: Story = {
  args: {
    children: 'echo hello world',
    textToCopy: 'echo hello world',
  },
};

export const WithoutTextToCopy: Story = {
  args: {
    children: 'echo hello world',
  },
};

export const WithTextToCopyJSX: Story = {
  args: {
    children: (
      <span>
        <strong>$</strong>echo hello world
      </span>
    ),
    textToCopy: 'echo hello world',
  },
};

export default { component: ShellBox } as Meta;
