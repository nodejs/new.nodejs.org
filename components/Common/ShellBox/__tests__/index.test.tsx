import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { IntlProvider } from 'react-intl';
import ShellBox from '../index';

const mockWriteText = jest.fn();
const originalNavigator = { ...window.navigator };

describe('ShellBox', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'navigator', {
      value: {
        clipboard: {
          writeText: mockWriteText,
        },
      },
    });
  });

  afterEach(() => {
    Object.defineProperty(window, 'navigator', {
      value: originalNavigator,
    });
  });

  it('should call clipboard API with `test` once', async () => {
    const user = userEvent.setup();
    const navigatorClipboardWriteTextSpy = jest
      .fn()
      .mockImplementation(() => Promise.resolve());

    Object.defineProperty(window.navigator, 'clipboard', {
      writable: true,
      value: {
        writeText: navigatorClipboardWriteTextSpy,
      },
    });

    render(
      <IntlProvider locale="en" onError={() => {}}>
        <ShellBox textToCopy="test">test</ShellBox>
      </IntlProvider>
    );
    const button = screen.getByRole('button');
    await user.click(button);
    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledTimes(1);
    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledWith('test');
  });
});
