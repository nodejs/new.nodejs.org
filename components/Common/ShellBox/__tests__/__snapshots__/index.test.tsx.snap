// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShellBox should render 1`] = `
<div>
  <pre
    class="shellBox"
  >
    <div
      class="top"
    >
      <span>
        SHELL
      </span>
      <button
        type="button"
      >
        components.common.shellBox.copy
      </button>
    </div>
    <code>
      test
    </code>
  </pre>
</div>
`;
